File: solution/1915/가장 큰 정사각형.py
f=open('./solution/1915/input.txt', 'r')
input=f.readline

n, m = map(int, input().split())
arr, dp = [[0]*(m+1)], [[0]*(m+1)]
for i in range(n):
    inp=[0]+[1 if c=='1' else 0 for c in input().rstrip()]
    arr.append(inp)
    dp.append(inp)

def chk(r, c):
    global dp
    a = dp[r-1][c-1]
    b = dp[r][c-1]
    c = dp[r-1][c]
    return min(a, b, c)+1

answer=0
for i in range(1, n+1):
    for j in range(1, m+1):
        if arr[i][j]:
            dp[i][j]=chk(i, j)
            answer=max(answer, dp[i][j])

print(answer*answer)

for row in dp:
    print(row)

File: solution/1915/가장 큰 정사각형.java
import java.io.*;
import java.util.*;

public class Main {

  static int N, M;
  static int[][] arr;

    public static void main(String[] args) throws Exception {
//    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream("./input.txt")));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[N+1][M+1];
        int answer=0;
        for (int i = 1; i <= N; i++) {
        	String str = br.readLine();
			for (int j = 1; j <= M; j++) {
				if (str.charAt(j-1)=='1') {
					arr[i][j] = Math.min(arr[i-1][j], Math.min(arr[i][j-1], arr[i-1][j-1])) + 1;
					answer = Math.max(answer, arr[i][j]);
				}
			}
		}
        bw.write(answer*answer+"\n");

        bw.flush();
        bw.close();
        br.close();
    }
}